Fixes to allow us to list the ara definitions

dleblanc opened the pull request:
This changes some ara definitions stuff to allow listing of ara definitions.

Specifically:
 - using the proper collection key
 - removing the individual key
 - using properties from both version and definition when building an AraDefinition

Ken's e2e test exercises this code path.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [ ] I have added tests for 100% coverage
- [ ] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [ ] I have bumped the version in setup.py

kroenlein commented on src/citrine/resources/ara_definition.py:
```diff
@@ -41,6 +41,7 @@ class AraDefinition(Resource["AraDefinition"]):
 
     @staticmethod
     def _get_dups(lst: List) -> List:
+        # Hmmn, this looks like a potentially costly operation?!
         return [x for x in lst if lst.count(x) > 1]
```
```suggestion
    def _get_dups(lst: List) -> List:
        seen = set()
        dups = [x for x in lst if x in seen or seen.add(x)]  # seen.add returns None
        return dups
```
But perhaps this is an off-topic optimizaton

dleblanc replied:
I agree that's a better approach. I'm going to leave it out of the PR for now just to keep focused.

kroenlein commented on src/citrine/resources/ara_definition.py:
```diff
@@ -221,10 +225,10 @@ def get_with_version(self, definition_uid: Union[UUID, str],
-        defn.definition_uid = data['definition_id']
-        defn.version_number = data['version_number']
-        defn.version_uid = data['id']
+        defn = AraDefinition.build(data['version']['ara_definition'])
```
Would it make sense for `version` to be a constant, given the number of times it appears?  I was originally thinking just use `_individual_key`, but that got cleared in this PR.

dleblanc replied:
I'll just extract those parts of the dict - it's only in this scope, and used 3 times.

dleblanc commented:
> Makes sense if it clears tests. Did you identify why we were incorrectly getting clean bills of health?

All the interactions are faked out in the unit tests, and they were representing a different JSON schema than what actually gets returned by the real API. In addition they were doing some decode->re-encode logic that doesn't work given that the AraDefinition munges bits of information from both the Ara 'definition', the 'version', as well as the 'ara_definition' JSON blob. That's not great, but a larger issue and not something I intend to fix here.

I've fixed them to represent the schema properly, and to not bother with the decode/re-encode business. I've also removed some overspecification in the tests and added some relevant asserts on the returned ara definitions.

dleblanc commented:
Rebuild!

dleblanc commented:
Rebuild!

kroenlein commented on tests/resources/test_ara_definition.py:
```diff
@@ -241,81 +243,89 @@ def test_add_all_ingredients(session, project):
-        path="projects/{}/ara-definitions".format(project_id),
-        json={"definition": collection.build(ara_definition).dump()}
-    )
-    assert session.last_call == expect_call
```
Good clean-up
